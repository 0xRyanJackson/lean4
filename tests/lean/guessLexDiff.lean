set_option showInferredTerminationBy true

def countUp (n i acc : Nat) : Nat :=
  if i < n then
    countUp n (i+1) (acc + i)
  else
    acc

def all42 (xs : Array Nat) (i : Nat) : Bool :=
  if h : i < xs.size then
    if xs[i] = 42 then
      all42 xs (i+1)
    else
      false
  else
    true

def henrik1 (xs : Array Nat) (i : Nat) : Bool :=
  if h : i < xs.size then
    if xs[i] = 42 then
      henrik1 (xs.push 42) (i+2)
    else
      false
  else
    true

/--
This checks
* the presentation of complex measures in the table
* that multiple recursive calls do not lead to the same argument tried multiple times.
* that we do not get measures from refined arguments
-/
def failure (xs : Array Nat) (i : Nat) : Bool :=
  if h : i < xs.size then
    failure xs i && failure xs i && failure xs (i + 1)
  else
  if h : i + 1 < xs.size then
    failure xs i
  else
  let j := i
  if h : j < xs.size then
    failure xs (j+1)
  else
  match i with
  | 0 => true
  | i+1 =>
      if h : i < xs.size + 5 then
        failure xs i
      else
        false

mutual
def mutual_failure (xs : Array Nat) (i : Nat) : Bool :=
  if h : i < xs.size then
    mutual_failure2 xs i && mutual_failure2 xs i && mutual_failure2 xs (i + 1)
  else
  if h : i + 1 < xs.size then
    mutual_failure2 xs i
  else
  let j := i
  if h : j < xs.size then
    mutual_failure2 xs (j+1)
  else
  match i with
  | 0 => true
  | i+1 =>
      if h : i < xs.size then
        mutual_failure2 xs i
      else
        false

def mutual_failure2 (xs : Array Nat) (i : Nat) : Bool :=
  if h : i < xs.size then
    mutual_failure xs i && mutual_failure xs i && mutual_failure xs (i + 1)
  else
  let j := i
  if h : j < xs.size then
    mutual_failure xs (j+1)
  else
  match i with
  | 0 => true
  | i+1 =>
      if h : i < xs.size then
        mutual_failure xs i
      else
        false
end
