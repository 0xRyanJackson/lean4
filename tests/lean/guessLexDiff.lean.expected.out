Inferred termination argument:
termination_by n - i
Inferred termination argument:
termination_by Array.size xs - i
Inferred termination argument:
termination_by Array.size xs - i
guessLexDiff.lean:35:4-35:16: error: fail to show termination for
  failure
with errors
argument #2 was not used for structural recursion
  failed to eliminate recursive application
    _root_.failure xs i

structural recursion cannot be used

Could not find a decreasing measure.
The arguments relate at each recursive call as follows:
(<, ≤, =: relation proved, ? all proofs failed, _: no proof attempted)
            i #1 #2
1) 35:4-16  =  =  =
2) 35:36-54 ?  <  ≤
3) 35:4-16  _  =  =
4) 42:4-20  _  <  <
5) 48:8-20  _  ?  ≤

#1: Array.size xs - i
#2: Array.size xs - (i + 1)

Please use `termination_by` to specify a decreasing measure.
guessLexDiff.lean:53:4-53:18: error: fail to show termination for
  mutual_failure
  mutual_failure2
with errors
structural recursion does not handle mutually recursive functions

Could not find a decreasing measure.
The arguments relate at each recursive call as follows:
(<, ≤, =: relation proved, ? all proofs failed, _: no proof attempted)
Call from mutual_failure to mutual_failure2 at 55:4-24:
   i #1 #2
i  =  ?  ?
#3 ?  =  ?
Call from mutual_failure to mutual_failure2 at 55:52-78:
   i #1 #2
i  ?  _  _
#3 _  <  _
Call from mutual_failure to mutual_failure2 at 55:4-24:
   i #1 #2
i  _  _  _
#3 _  =  _
Call from mutual_failure to mutual_failure2 at 62:4-28:
   i #1 #2
i  _  _  _
#3 _  <  _
Call from mutual_failure to mutual_failure2 at 68:8-28:
   i #1 #2
i  _  _  _
#3 _  ?  _
Call from mutual_failure2 to mutual_failure at 74:4-23:
   i #3
i  _  _
#1 _  _
#2 _  _
Call from mutual_failure2 to mutual_failure at 74:50-75:
   i #3
i  _  _
#1 _  _
#2 _  _
Call from mutual_failure2 to mutual_failure at 78:4-27:
   i #3
i  _  _
#1 _  _
#2 _  _
Call from mutual_failure2 to mutual_failure at 84:8-27:
   i #3
i  _  _
#1 _  _
#2 _  _


#1: Array.size xs - i
#2: Array.size xs - (i + 1)
#3: Array.size xs - i

Please use `termination_by` to specify a decreasing measure.
